### PR TITLE
For C-like targets, emit resource declarations before other globals

### PR DESCRIPTION
**Pull request by saipraveenb25**
This is to fix a weird edge case in the DXC compiler, where using a array of `SamplerState` as a function parameter type before declaring any resources of the same type leads to an 'incomplete type' error.

To fix this, we emit resource declarations & bindings before all other globals (except for types that the resource declarations depend on)

### CODE REVIEW

File: tests/autodiff/intellisense-test.slang [hunk 1]
```diff
@@ -0,0 +1,39 @@
+RWStructuredBuffer<float> outputBuffer;
```
Comment (csyonghe) [inline tests/autodiff/intellisense-test.slang:1]: If this is intended to test language server, use `//TEST:LANG_SERVER:` and place it under `tests/language-server`, you can refer to other tests there to see how to test the language server features.

File: tests/autodiff/multi-differential.slang [hunk 2]
```diff
@@ -0,0 +1,96 @@
+//TEST_IGNORE_FILE:
```
Comment (csyonghe) [inline tests/autodiff/multi-differential.slang]: Why ignored?

File: tests/autodiff/higher-order-custom-derivative.slang [hunk 3]
```diff
@@ -0,0 +1,39 @@
+//TEST_IGNORE_FILE:
```
Comment (csyonghe) [inline tests/autodiff/higher-order-custom-derivative.slang]: Why ignored?

File: tests/autodiff/multi-differential.slang (hunk 2 above)
Comment (saipraveenb25) [reply tests/autodiff/multi-differential.slang]: Whoops, all these tests were experimental.. (accidentally committed)
I've removed them all now

File: tests/autodiff/swizzled-store.slang [hunk 4]
```diff
@@ -0,0 +1,27 @@
+//TEST_IGNORE_FILE:
```
Comment (csyonghe) [inline tests/autodiff/swizzled-store.slang]: delete if it is not testing anything?

File: tests/autodiff/higher-order-custom-derivative.slang (hunk 3 above)
Comment (saipraveenb25) [reply tests/autodiff/higher-order-custom-derivative.slang]: Whoops, all these tests were experimental.. (accidentally committed)
I've removed them all now

File: tests/autodiff/swizzled-store.slang (hunk 4 above)
Comment (saipraveenb25) [reply tests/autodiff/swizzled-store.slang]: Accidentally committed test

**Pull request merged**